java-checks/src/main/java/org/sonar/java/checks/xml/maven/DeprecatedPomPropertiesCheck.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.xml.maven;
+
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.tag.Tag;
+import org.sonar.java.xml.XPathXmlCheck;
+import org.sonar.java.xml.XmlCheckContext;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+import org.w3c.dom.Node;
+
+import javax.xml.xpath.XPathExpression;
+
+@Rule(
+  key = "S3421",
+  name = "Deprecated \"${pom}\" properties should not be used",
+  priority = Priority.MAJOR,
+  tags = {Tag.OBSOLETE, Tag.MAVEN})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LANGUAGE_RELATED_PORTABILITY)
+@ActivatedByDefault
+@SqaleConstantRemediation("2min")
+public class DeprecatedPomPropertiesCheck extends XPathXmlCheck {
+  private static final String POM_PROPERTY_PREFIX = "${pom.";
+  private static final String POM_PROPERTY_SUFFIX = "}";
+  private XPathExpression textsExpression;
+
+  @Override
+  public void precompileXPathExpressions(XmlCheckContext context) {
+    textsExpression = context.compile("//*[text()]");
+  }
+
+  @Override
+  public void scanFileWithXPathExpressions(XmlCheckContext context) {
+    if ("pom.xml".equals(context.getFile().getName())) {
+      for (Node textNode : context.evaluateOnDocument(textsExpression)) {
+        String text = textNode.getFirstChild().getNodeValue();
+        while (StringUtils.contains(text, POM_PROPERTY_PREFIX)) {
+          String property = extractPropertyName(text);
+          reportIssue(textNode, "Replace \"pom." + property + "\" with \"project." + property + "\".");
+          text = skipFirstProperty(text);
+        }
+      }
+    }
+  }
+
+  private static String skipFirstProperty(String text) {
+    String newText = text.substring(text.indexOf(POM_PROPERTY_PREFIX));
```

Comment:
no need to do two substring, you should search for index of suffix starting at index of prefix